Name=Nisarg Patel
student id-8878184

1.Utilize the diabetes dataset from lab 4. Perform cross-validation on nine polynomial models, ranging from degree 0 to 8.

In [9]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import cross_validate
from sklearn.pipeline import Pipeline
from sklearn.metrics import make_scorer, mean_absolute_error , mean_absolute_percentage_error , r2_score

# Load the diabetes dataset
X, y = load_diabetes(return_X_y=True, as_frame=True)

X.head(10)



age       sex       bmi        bp        s1        s2        s3  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005671 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   
5 -0.092695 -0.044642 -0.040696 -0.019442 -0.068991 -0.079288  0.041277   
6 -0.045472  0.050680 -0.047163 -0.015999 -0.040096 -0.024800  0.000779   
7  0.063504  0.050680 -0.001895  0.066630  0.090620  0.108914  0.022869   
8  0.041708  0.050680  0.061696 -0.040099 -0.013953  0.006202 -0.028674   
9 -0.070900 -0.044642  0.039062 -0.033214 -0.012577 -0.034508 -0.024993   

         s4        s5        s6  
0 -0.002592  0.019908 -0.017646  
1 -0.039493 -0.068330 -0.092204  
2 -0.002592  0.002864 -0.025930  
3  0.034309  0.022692 -0.009362  
4 -0.002592 -0.031991 -0.046641  
5 -0.076395 -0.041180 -0.096346  
6 -0.039493 -0.062913 -0.038357  
7  0.017703 -0.035817  0.003064  
8 -0.002592 -0.014956  0.011349  
9 -0.002592  0.067736 -0.013504

In [11]:
X.shape , y.shape
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=100)
X.shape , y.shape

((442, 10), (442,))

cross-validation on nine polynomial models

In [12]:
def makeModelPipe(d,bias=False):
    model = Pipeline([('polynomial', PolynomialFeatures(degree=d , include_bias=bias)), 
                    ('linear', LinearRegression())])
    return model

In [13]:
model = {}
scores = {}
_scoring = {
        'r2': make_scorer(r2_score),
        'mae': make_scorer(mean_absolute_error),
        'mape': make_scorer(mean_absolute_percentage_error)
    } # Scores 
for i in range(9): # 9 models with 0-8 degrees
    if(i == 0):
        model[i] = makeModelPipe(i,True)
    else:
        model[i] = makeModelPipe(i)
    scores[i] = cross_validate(model[i], X_train, y_train, 
                               cv=10, #10 cross-validation splits for each model
                               scoring = _scoring, 
                               n_jobs=-1, 
                               verbose=2)

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 16 concurrent workers.
[Parallel(n_jobs=-1)]: Done   3 out of  10 | elapsed:    3.7s remaining:    8.7s
[Parallel(n_jobs=-1)]: Done  10 out of  10 | elapsed:    3.7s finished
C:\Users\nisar\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
10 fits failed out of a total of 10.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
3 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\nisar\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\nisar\anaconda3\lib\site-packages\sklearn\pipeline.py", line 390,

2.Construct a table summarizing the cross-validation results. Each model should have a separate row in the table. Include the R-Squared and Mean Absolute Error (MAE) metrics for each model. Calculate the mean value and standard deviation of these metrics from the cross-validation. Include both values.

In [14]:
scores_m_std = pd.DataFrame(scores)
scores_m_std

0  \
fit_time    [0.004000425338745117, 0.0030019283294677734, ...   
score_time  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   
test_r2     [nan, nan, nan, nan, nan, nan, nan, nan, nan, ...   
test_mae    [nan, nan, nan, nan, nan, nan, nan, nan, nan, ...   
test_mape   [nan, nan, nan, nan, nan, nan, nan, nan, nan, ...   

                                                            1  \
fit_time    [0.009999513626098633, 0.00800013542175293, 0....   
score_time  [0.002000093460083008, 0.0029993057250976562, ...   
test_r2     [0.5160554356206847, 0.4961199892544894, 0.502...   
test_mae    [43.14255343978075, 39.09350815099935, 42.2150...   
test_mape   [0.4543257558012658, 0.3490880986986697, 0.412...   

                                                            2  \
fit_time    [0.013533830642700195, 0.012533903121948242, 0...   
score_time  [0.0030002593994140625, 0.0030002593994140625,...   
test_r2     [0.2891907202379461, 0.47232521483450507, 0.15...   
test_mae    [51.13618562839776, 38.83560737850228, 52.2485...   
test_mape   [0.47437084508641436, 0.3401708444143411, 0.51...   

                                                            3  \
fit_time    [0.015982866287231445, 0.016002416610717773, 0...   
score_time  [0.00400233268737793, 0.003000497817993164, 0....   
test_r2     [-379.5317484781913, -532.9911675373618, -21.3...   
test_mae    [376.9301825859988, 583.7592600993934, 224.643...   
test_mape   [2.433991723218688, 5.01045871372179, 2.377602...   

                                                            4  \
fit_time    [0.04640960693359375, 0.040024757385253906, 0....   
score_time  [0.003002643585205078, 0.0020055770874023438, ...   
test_r2     [-33.090869251207266, -88.31095846434829, -138...   
test_mae    [292.14773749786696, 362.1108800814603, 458.58...   
test_mape   [3.4144564120213414, 3.451151452298105, 3.9795...   

                                                            5  \
fit_time    [0.13153815269470215, 0.14252519607543945, 0.1...   
score_time  [0.0020012855529785156, 0.002106904983520508, ...   
test_r2     [-29.592896187886666, -86.61773032624531, -119...   
test_mae    [280.2846283202462, 359.13669671768247, 429.87...   
test_mape   [3.331969842255598, 3.4030332973763757, 3.7733...   

                                                            6  \
fit_time    [0.3480796813964844, 0.3271760940551758, 0.327...   
score_time  [0.003996849060058594, 0.006009101867675781, 0...   
test_r2     [-29.58785008719539, -86.58227968061485, -119....   
test_mae    [280.2811242420301, 359.15903157872714, 429.85...   
test_mape   [3.331557987914975, 3.4029361756103653, 3.7734...   

                                                            7  \
fit_time    [0.7247567176818848, 0.8353455066680908, 0.830...   
score_time  [0.009017229080200195, 0.0070035457611083984, ...   
test_r2     [-29.588171182232024, -86.581691423591, -119.2...   
test_mae    [280.28502654025993, 359.1595189132623, 429.84...   
test_mape   [3.331595769474456, 3.4029319050778795, 3.7734...   

                                                            8  
fit_time    [2.3417022228240967, 2.2040469646453857, 2.380...  
score_time  [0.026006698608398438, 0.025984525680541992, 0...  
test_r2     [-29.58822483600056, -86.58164185822442, -119....  
test_mae    [280.28534674536843, 359.1595409617914, 429.84...  
test_mape   [3.331597425119039, 3.402932159537936, 3.77343...

In [16]:
scores_copy = scores_m_std.copy() 

for c,col in enumerate(scores_copy):
    for idx,row in enumerate(scores_copy[col]):
        scores_copy.iloc[idx,c] = [scores_copy.iloc[idx,c].mean(),scores_copy.iloc[idx,c].std()]

        
# mean and the standard deviation of each model        

final_scores = scores_copy.T
final_scores = final_scores[['test_r2','test_mae','test_mape']]
final_scores

test_r2  \
0                                 [nan, nan]   
1  [0.4621091446405231, 0.12142302506692582]   
2  [0.3590063236887825, 0.16263731909826237]   
3    [-394.9830393092353, 659.6237824384839]   
4  [-43.858871449998205, 37.540368668775635]   
5  [-40.648848774800896, 33.250482527055915]   
6    [-40.63320088307083, 33.23972167986735]   
7    [-40.63306493390254, 33.23887303584974]   
8   [-40.63305941384206, 33.238869898137686]   

                                   test_mae  \
0                                [nan, nan]   
1   [45.520703028232376, 5.043739127371635]   
2    [48.35514905583644, 6.160591413193763]   
3  [366.71227122108974, 211.19415540796197]   
4    [298.3361555331322, 71.69112471407236]   
5    [289.98747863863207, 67.1632446692545]   
6     [289.955854531642, 67.16010947868479]   
7    [289.95577741339275, 67.1596380593687]   
8    [289.9557885564982, 67.15962643086806]   

                                   test_mape  
0                                 [nan, nan]  
1  [0.4011607382603777, 0.05749216969647432]  
2  [0.4164525161173997, 0.06478405238247109]  
3   [2.9658757229582173, 1.9160435366988113]  
4    [2.513815795018931, 0.7864631393118221]  
5   [2.4667288087921984, 0.7491464198466906]  
6    [2.466472097428913, 0.7491859441345341]  
7    [2.466485058929671, 0.7491823484072462]  
8   [2.4664850354788763, 0.7491825080283228]

In [17]:
final_scores_2 = pd.DataFrame()
for c,col in enumerate(final_scores):
    for idx,row in enumerate(final_scores[col]):
        # print(idx,row[0],row[1])
        final_scores_2.loc[idx,f'{col}_mean'] = row[0]
        final_scores_2.loc[idx,f'{col}_std'] = row[1]
        
        
# mean and std as different columns
final_scores_2

test_r2_mean  test_r2_std  test_mae_mean  test_mae_std  test_mape_mean  \
0           NaN          NaN            NaN           NaN             NaN   
1      0.462109     0.121423      45.520703      5.043739        0.401161   
2      0.359006     0.162637      48.355149      6.160591        0.416453   
3   -394.983039   659.623782     366.712271    211.194155        2.965876   
4    -43.858871    37.540369     298.336156     71.691125        2.513816   
5    -40.648849    33.250483     289.987479     67.163245        2.466729   
6    -40.633201    33.239722     289.955855     67.160109        2.466472   
7    -40.633065    33.238873     289.955777     67.159638        2.466485   
8    -40.633059    33.238870     289.955789     67.159626        2.466485   

   test_mape_std  
0            NaN  
1       0.057492  
2       0.064784  
3       1.916044  
4       0.786463  
5       0.749146  
6       0.749186  
7       0.749182  
8       0.749183

In [18]:
final_scores_2.index.name = 'Polynomial Degree'
final_scores_2

test_r2_mean  test_r2_std  test_mae_mean  test_mae_std  \
Polynomial Degree                                                           
0                           NaN          NaN            NaN           NaN   
1                      0.462109     0.121423      45.520703      5.043739   
2                      0.359006     0.162637      48.355149      6.160591   
3                   -394.983039   659.623782     366.712271    211.194155   
4                    -43.858871    37.540369     298.336156     71.691125   
5                    -40.648849    33.250483     289.987479     67.163245   
6                    -40.633201    33.239722     289.955855     67.160109   
7                    -40.633065    33.238873     289.955777     67.159638   
8                    -40.633059    33.238870     289.955789     67.159626   

                   test_mape_mean  test_mape_std  
Polynomial Degree                                 
0                             NaN            NaN  
1                        0.401161       0.057492  
2                        0.416453       0.064784  
3                        2.965876       1.916044  
4                        2.513816       0.786463  
5                        2.466729       0.749146  
6                        2.466472       0.749186  
7                        2.466485       0.749182  
8                        2.466485       0.749183

3.dentification of the Best Model: Identify the model that exhibits the highest performance based on the R-Squared and MAE metrics. Provide an explanation for choosing this specific model.

-Choosing the degree 2 multivariate linear regression model above the other models due to its higher R-squared value (0.4621), which shows that it explains more variance in the target variable.

The Multivariate Linear Regression has the lowest MAPE (0.40), which shows that, generally speaking, its predictions are the closest to the actual values.

The Multi Variant Linear Regression model with degree-2 finds a balance between identifying underlying trends in the data (high R-squared) and making precise predictions (relatively low MAE), hence I suggest using it after analysing the findings.